enhancements/installer/agent-based-installation-in-hive.md

```diff
@@ -0,0 +1,539 @@
+kubernetes-native it is more natural for it to be separate.
+
+The discovery ISO can be downloaded from a URL that is available in the
+resource's Status.
+
+The details of this resource definition were discussed [in a
+pull request](https://github.com/openshift/assisted-service/pull/969).
+
+Summary of fields:
+
+Spec
+* egress proxy settings
+* a list of NTP sources (hostname or IP) to be added to all cluster
+hosts. They are added to any NTP sources that were configured through other
+means.
+* a list of SSH public keys that will be added to all agents for use in debugging.
+* pull secret reference
+* a label selector for Agents. This is how Agent resources are identified as
+belonging to a particular InstallEnv.
+* labels to add to Agent resources upon creation. These labels should match
+the label selector above.
+* ClusterDeployment reference. This is a temporary field that associated an
+InstallEnv with a single cluster definition. This is required because the
+backend assisted service is not capable of producing a discovery ISO unless it
+is scoped to a cluster. That limitation will be removed in future releases.
+
+Status
+* a URL to download the discovery ISO produced for this InstallEnv.
+
+**Agent**
+Agent is a new resource, part of the assisted installer's new operator, that
+represents a host that is destined to become part of an OpenShift cluster, and
+is running an agent that is able to run inspection and installation tasks. It
+correlates to the "host" resource in the Assisted Installer's REST API.
+
+The details of this API are available [as
+code](https://github.com/openshift/assisted-service/pull/861), with some
+details yet to be changed.
+
+Summary of fields:
+
+Spec
+* Role that the Node will have
+* Hostname
+* Approved
+* Install device, set automatically when there is platform-integrated
+automation booting the discovery ISO, or otherwise set by the user in a
+boot-it-yourself scenario.
+
+Status
+* Comprehensive inspection and validation data
+
+#### Hive Integration
+
+Hive has a [ClusterDeployment
+CRD](https://github.com/openshift/hive/blob/master/docs/using-hive.md#clusterdeployment)
+resource that represents a cluster to be created. It includes a "platform"
+section where platform-specific details can be captured. For an agent-based
+workflow, this section will include a new `AgentBareMetal` platform that
+contains such fields as:
+
+* API VIP
+* API VIP DNS name
+* IngressVIP
+
+A new `InstallStrategy` section of the Spec enables the API to describe a way
+of installing a cluster other than the default of using `openshift-install`.
+The new field has an "agent" option that would be common to any Agent-driven
+installation, regardless of platform. Fields include:
+
+* AgentSelector, a label selector for identifying which Agent resources should
+be included in the cluster.
+* ProvisionRequirements, where the API user can specify how many agents to
```

Comment:
My hope was that in the future, we use Hive MachinePools which map out to MachineSets (as the installer does), after which an as yet vaporware agent/metal machine provider would start finding Agents to fulfill the Machine requests that resulted from the MachineSets.

ie. MachinePool -> MachineSet -> Machine -> Agent

We have not really gotten into MachinePool support thus far. Would this be the missing piece you're referring to?

Reply:
Yeah, I guess this is how the "Machine resources on the hub cluster" work will eventually intersect with Hive agent-based provisioning

And we're taking an interim approach here so that we can make progress without depending on "machine resources on the hub cluster"

If so, I'd like to see some discussion of this future direction and how we ourselves transitioning to a MachinePool based add-worker API from the proposed API, and whether/how we'll retain backwards compatibility over that transition.

Reply:
https://issues.redhat.com/browse/HIVE-1393

Reply:
Given functional central machine management, and an as yet non-existant BMH/Agent mapping machine controller, we expect the very rough flow to be a hive MachinePool creating MachineSets, Machines being created automatically, this new BMH->Agent machine controller finding Agents to assign to satisfy the request for each Machine.

I don't see us being able to do this backward compatible. It feels like in 4.8 we'd need people to be creating MachinePools that don't get used at all, and then after an upgrade suddenly they are. This feels dangerous to me as it can involve Machine's getting nuked by virtue of an unexpected change rolling out to MachineSets. You could likely take a pre-existing cluster with all it's worker nodes, and then add new MachinePools resulting in new pools of worker nodes, but it feels risky to try to later inject a MachinePool to "adopt" the pre-existing 4.8 Machines.

Reply:
A new section `#### Future Use of MachinePools and Machine API` discusses these issues below.